File: 402-remove-k-digits/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        string stack = "";  // Using string as stack
        
        for(char ch : num){
            while(!stack.empty() && k > 0 && stack.back() > ch){
                stack.pop_back();
                k--;
            }
            stack.push_back(ch);
        }
        
        // if not able to remove all k-digits 
        // remove the digits from the end of stack coz its monotonic increasing
        while(k > 0 && !stack.empty()){
            stack.pop_back();
            k--;
        }
        
        // remove the leading zeroes
        int index = 0;
        while(index < stack.length() && stack[index] == '0'){
            index++;
        }
        string result = stack.substr(index);
        return result.empty() ? "0" : result;
    }
};

File: 402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) 
    {
        Deque<Character> stack = new ArrayDeque<>();
        for(int i = 0; i < num.length(); i++){
            char ch = num.charAt(i);
            while(!stack.isEmpty() && k > 0 && stack.peek() > ch){
                stack.pop();
                k--;
            }
            stack.push(ch);
        }
        
        // if not able to remove all k-digits 
        // remove the digits from the top of stack coz its monotonic inscreasing
        while(k > 0 && !stack.isEmpty()){
            stack.pop();
            k--;
        }
        
        StringBuilder sb = new StringBuilder();
        while(!stack.isEmpty()){
            sb.append(stack.pop());
        }
        sb.reverse();
        
        int index = 0;
        // remove the leading zeroes
        while(index < sb.length() && sb.charAt(index) == '0'){
            index++;
        }
        String result = sb.substring(index);
        return result.isEmpty() ? "0" : result;
    }
}